pkg/tasktesting/logs.go

```diff
@@ -1,70 +1,137 @@
-// from https://github.com/tektoncd/cli/blob/c996b3004650658c73ae8b1c0ce98f5165107af5/test/framework/logs.go
 package tasktesting
 
 import (
 	"context"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"log"
-	"strings"
 	"time"
 
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/kubernetes"
-	"knative.dev/pkg/test/logging"
 )
 
-// CollectPodLogs will get the logs for all containers in a Pod
-func CollectPodLogs(c *kubernetes.Clientset, podName, namespace string, logf logging.FormatLogger, podEventsDone chan<- bool) {
-	logs, err := getContainerLogsFromPod(c, podName, namespace, podEventsDone)
-	if err != nil {
-		logf("Could not get logs for pod %s: %s", podName, err)
+// getEventsAndLogsOfPod streams events of the pod until all containers are ready,
+// and streams logs for each container once ready. It stops if there are any
+// sends on the errs channels or if the passed context is cancelled.
+func getEventsAndLogsOfPod(
+	ctx context.Context,
+	c kubernetes.Interface,
+	pod *corev1.Pod,
+	errs chan error) {
+	quitEvents := make(chan bool)
+	podName := pod.Name
+	podNamespace := pod.Namespace
+
+	go watchPodEvents(
+		ctx,
+		c,
+		podName,
+		podNamespace,
+		quitEvents,
+		errs,
+	)
+
+	watchingEvents := true
+	for _, container := range pod.Spec.Containers {
+		err := streamContainerLogs(ctx, c, podNamespace, podName, container.Name)
```

Comment:
Assuming a TaskRun is composed of more than one container, in what order are we iterating over them? How do we know we stream the logs of the first container that the TaskRun spins up?

Reply:
`.Spec.Containers` is a slice and therefore ordered. My assumption is that the Tekton steps (which are also a slice) are moved to the pod spec in the same order. I have not verified this in the codebase, but in all the test runs I've seen so far this holds true. So I'd say we go with this until proven otherwise :)